File: Online Judges/HackerEarth/Min Cost Flow/AmazingRace.cpp
#include <bits/stdc++.h>
#define lli long long int
using namespace std;
const int maxSize = 200 + 200 + 2 + 1;
int n, m, k; lli inf = 1e18;
lli matrixGraph[maxSize][maxSize][2];
// source -> team (n) -> check (m) -> target

void read(int a[][2], int n)
{
  for (int i = 0; i < n; i ++) scanf("%d %d", &a[i][0], &a[i][1]);
}

void addEdge(vector<int> graph[], int u, int v, int f, int c)
{
  graph[u].push_back(v);
  matrixGraph[u][v][0] = f, matrixGraph[u][v][1] = c;
  graph[v].push_back(u);
  matrixGraph[v][u][0] = 0, matrixGraph[v][u][1] = -c;
}

double distance(int a[], int b[], double s)
{
  double x = a[0], y = a[1], p = b[0], q = b[1];
  return(ceil(((p-x)*(p-x) + (q-y)*(q-y)) / (s * s)));
}

int bellmannFord(vector<int> graph[], int prev[], int source, int target)
{
  lli cost[maxSize]; for (int i = 0; i < maxSize; i ++) cost[i] = inf;
  cost[source] = 0;
  for (int i = 0; i < n + m + 2 + 1 - 1; i ++)
  {
    int done = 0;
    for (int u = 0; u < n + m + 2 + 1; u ++)
      for (auto v: graph[u])
        if (matrixGraph[u][v][0] && max(cost[u], matrixGraph[u][v][1]) < cost[v])
        {
          cost[v] = max(cost[u], matrixGraph[u][v][1]);
          prev[v] = u;
          done = 1;
        }
    if (!done) break;
  }
  return(cost[target] != inf);
}

void minCostFlow(vector<int> graph[], lli *maxFlow, lli *minCost, int source, int target)
{
  int prev[maxSize];
  while (bellmannFord(graph, prev, source, target))
  {
    int v = target; lli flow = inf, cost = 0;
    while (v != source)
    {
      flow = min(flow, matrixGraph[prev[v]][v][0]);
      cost = max(cost, matrixGraph[prev[v]][v][1]);
      v = prev[v];
    }

    *maxFlow += flow;
    *minCost = max(*minCost, cost);
    v = target;
    while (v != source)
    {
      matrixGraph[prev[v]][v][0] -= flow;
      matrixGraph[v][prev[v]][0] += flow;
      v = prev[v];
    }
  }
}

int main()
{
  memset(matrixGraph, 0, sizeof(matrixGraph));
  scanf("%d %d %d", &n, &m, &k);
  int teamPos[n][2], checkPos[m][2], teamSpeed[n];
  read(teamPos, n); read(checkPos, m);
  for (int i = 0; i < n; i ++) scanf("%d", &teamSpeed[i]);

  vector<int> graph[maxSize];
  for (int i = 0; i < n; i ++) addEdge(graph, 0, i + 1, 1, 0);
  for (int i = 0; i < m; i ++) addEdge(graph, i + n + 1, n + m + 1, 1, 0);
  for (int i = 0; i < n; i ++) for (int j = 0; j < m; j ++)
    addEdge(graph, i + 1, j + n + 1, 1, distance(teamPos[i], checkPos[j], teamSpeed[i]));
  addEdge(graph, n + m + 1, n + m + 2, k, 0);

  lli maxFlow = 0, minCost = 0; minCostFlow(graph, &maxFlow, &minCost, 0, m + n + 2);
  printf("%lld\n", minCost);

  return(0);
}

File: Online Judges/HackerEarth/Min Cost Flow/AmazingRace.py
from math import ceil
inf = 1e8414120412431

def printGraph(graph):
    for u in graph:
        print(u, end='')
        for v in graph[u]:
            print(" ->", (v, graph[u][v]), end='')
        print()

def distance(a, b, s):
    x, y, p, q = a[0], a[1], b[0], b[1]
    return(ceil(((p-x)**2 + (q-y)**2) / s**2))

def bellmannFord(graph, prev, source, target):
    cost = {}
    for i in graph: cost[i] = inf
    cost[source] = 0
    for i in range(len(graph) - 1):
        done = 0
        for u in graph:
            for v in graph[u]:
                if (graph[u][v][0] != 0 and max(cost[u], graph[u][v][1]) < cost[v]):
                    cost[v] = max(cost[u], graph[u][v][1])
                    prev[v] = u
                    done = 1
        if (done == 0): break
    return(cost[target] != inf)


def minCostFlow(graph, source, target):
    prev, maxFlow, minCost = {}, 0, 0
    while (bellmannFord(graph, prev, source, target)):
        v, flow, cost = target, inf, 0
        while (v != source):
            flow = min(flow, graph[prev[v]][v][0])
            cost = max(cost, graph[prev[v]][v][1])
            v = prev[v]
        maxFlow += flow
        minCost = max(minCost, cost)

        v = target
        while (v != source):
            graph[prev[v]][v][0] -= flow
            graph[v][prev[v]][0] += flow
            v = prev[v]
        prev = {}

    return(maxFlow, minCost)

n, m, k = map(int, input().split())
teamPos, team = [], []
for i in range(n):
    teamPos += [list(map(int, input().split()))]
    team += ["team %d" % i]
checkPos, check = [], []
for i in range(m):
    checkPos += [list(map(int, input().split()))]
    check += ["check %d" % i]
teamSpeed = list(map(int, input().split()))
# print(n, m, k, teamPos, teamSpeed, checkPos)

# [flow, cost]
graph = {}
graph["source"], graph["mid"], graph["target"] = {}, {}, {}
for i in range(m):
    graph[check[i]] = {}
    graph[check[i]]["mid"] = [1, 0]
    graph["mid"][check[i]] = [0, 0]
for i in range(n):
    graph["source"][team[i]] = [1, 0]
    graph[team[i]] = {}
    graph[team[i]]["source"] = [0, 0]
    for j in range(m):
        d = distance(teamPos[i], checkPos[j], teamSpeed[i])
        print(d)
        graph[team[i]][check[j]] = [1, d]
        graph[check[j]][team[i]] = [0, -d]
graph["mid"]["target"] = [k, 0]
graph["target"]["mid"] = [0, 0]

# printGraph(graph)
flow, cost = minCostFlow(graph, "source", "target")
# print(flow, cost)
print(cost)
